### PR TITLE
[FXML-1930] Implement mul folding

### PR DESCRIPTION
**Pull request by TinaAMD**
Implement folding for constant multiplications.

* Add tests
* Use unspecified integer overflow behavior

### CODE REVIEW

File: mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    assert(lhsElemType == rhsElemType);
+
+    auto resultType = mulOp.getType();
+    auto resultElementType = resultType.getElementType();
+    if (isa<IntegerType>(lhsElemType)) {
+      assert(isa<IntegerType>(rhsElemType) &&
+             isa<IntegerType>(resultElementType));
+      auto resultElementWidth = resultElementType.getIntOrFloatBitWidth();
+      assert(resultElementWidth >= lhsElemType.getIntOrFloatBitWidth() &&
+             "The multiplication is expected to have an at least a big  output "
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp:78]: `as big` and extra space

File: mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+             "as input type");
+      // TODO: To implement shifts > 0, capture the shift value stored in the
+      // mul here
+      bool mulOverflowed;
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp:82]: Initialize to false

File: mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    auto resultType = mulOp.getType();
+    auto resultElementType = resultType.getElementType();
+    if (isa<IntegerType>(lhsElemType)) {
+      assert(isa<IntegerType>(rhsElemType) &&
+             isa<IntegerType>(resultElementType));
+      auto resultElementWidth = resultElementType.getIntOrFloatBitWidth();
+      assert(resultElementWidth >= lhsElemType.getIntOrFloatBitWidth() &&
+             "The multiplication is expected to have an at least a big  output "
+             "as input type");
+      // TODO: To implement shifts > 0, capture the shift value stored in the
```
Comment (mgehre-amd) [inline mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp:80]: Does this comment add anything on top of the message "Non-zero shift folding is currently not implemented." above?

**Review by mgehre-amd — APPROVED** [review]
Looks good with minor comments to be addressed

File: mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp (hunk 3 above)
Comment (TinaAMD) [reply mlir/lib/Dialect/Tosa/Transforms/TosaFoldConstantMul.cpp]: I dropped it, was just a more detailed note on where one could implement it.

**Pull request merged**